google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddAdCustomizer.java

```diff
@@ -0,0 +1,498 @@
+    Feed adCustomizerFeed =
+        Feed.newBuilder()
+            .setName(StringValue.of(feedName))
+            .addAttributes(nameAttribute)
+            .addAttributes(priceAttribute)
+            .addAttributes(dateAttribute)
+        .build();
+
+    FeedOperation feedOperation =
+        FeedOperation.newBuilder()
+            .setCreate(adCustomizerFeed)
+        .build();
+
+    try (FeedServiceClient feedServiceClient =
+        googleAdsClient.getLatestVersion().createFeedServiceClient()) {
+
+      MutateFeedsResponse response =
+          feedServiceClient.mutateFeeds(Long.toString(customerId), ImmutableList.of(feedOperation));
+
+      String feedResourceName = response.getResults(0).getResourceName();
+      System.out.printf("Added feed with resource name %s.%n", feedResourceName);
+      return feedResourceName;
+    }
+  }
+
+  /**
+   * Retrieves all the attributes for a feed and returns them in a map using the attribute names as
+   * keys.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param feedResourceName the resource name of the feed.
+   * @return the attributes of the feed.
+   */
+  private Map<String, FeedAttribute> getFeedAttributes(
+      GoogleAdsClient googleAdsClient, long customerId, String feedResourceName) {
+    String query = String.format(
+        "SELECT feed.attributes, feed.name from feed where feed.resource_name = '%s'",
+        feedResourceName);
+
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(Long.toString(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .build();
+
+    Map<String, FeedAttribute> feedAttributes = new HashMap<>();
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      SearchPagedResponse searchPagedResponse = googleAdsServiceClient.search(request);
+
+      Feed feed = searchPagedResponse.iterateAll().iterator().next().getFeed();
+
+      System.out.printf(
+          "Found the following attributes for feed with name %s:%n",
+          feed.getName().getValue());
+      for(FeedAttribute feedAttribute : feed.getAttributesList()) {
+        System.out.printf(
+            "\t\"%s\" with id %d and type %s%n",
+            feedAttribute.getName().getValue(),
+            feedAttribute.getId().getValue(),
+            feedAttribute.getType());
+        feedAttributes.put(feedAttribute.getName().getValue(), feedAttribute);
+      }
+    }
+    return feedAttributes;
+  }
+
+  /**
+   * Creates a feed mapping to set the feed as an ad customizer feed.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param feedResourceName the resource name of the feed.
+   * @param feedAttributes the attributes of the feed.
+   */
+  private void createAdCustomizerMapping(
```

Comment:
Do ad customizers actually require a mapping? I thought you could use arbitrary feeds with customizers?

Reply:
Nope, creating customized ads without having added a mapping fails with

```
error_code {
  ad_error: MISSING_AD_CUSTOMIZER_MAPPING
}
message: "The feed referenced must have ad customizer mapping to be used in a customizer tag."
```